pkg/api/types.go

```diff
@@ -320,8 +320,81 @@ type Result struct {
+}
+
+// BuildStepInfo holds information about a particular step in the build process.
+type BuildStepInfo struct {
+	// FailureReason is a camel case reason that is used by the machine to reply
+	// back to the OpenShift builder with information why any of the steps in the
+	// build, failed.
+	FailureReason StepFailureReason
 }
 
+// StepFailureReason holds the type of failure that occured during the build
+// process.
+type StepFailureReason string
+
+const (
+	// ReasonAssembleFailed is the reason associated with the Assemble script
+	// failing.
+	ReasonAssembleFailed StepFailureReason = "AssembleFailed"
+
+	// ReasonPullBuilderImageFailed is the reason associated with failing to pull
+	// the builder image.
+	ReasonPullBuilderImageFailed StepFailureReason = "PullBuilderImageFailed"
+
+	// ReasonPullRuntimeImageFailed is the reason associated with failing to pull
+	// the runtime image.
+	ReasonPullRuntimeImageFailed StepFailureReason = "PullRuntimeImageFailed"
+
+	// ReasonCommitContainerFailed is the reason associated with failing to
+	// commit the container to the final image.
+	ReasonCommitContainerFailed StepFailureReason = "ContainerCommitFailed"
+
+	// ReasonFetchSourceFailed is the reason associated with failing to download
+	// the source of the build.
+	ReasonFetchSourceFailed StepFailureReason = "FetchSourceFailed"
+
+	// ReasonDockerImageBuildFailed is the reasons associated with a failed
+	// Docker image build.
+	ReasonDockerImageBuildFailed StepFailureReason = "DockerImageBuildFailed"
+
+	// ReasonDockerFileCreateFailed is the reason associated with failing to create a
+	// Dockerfile for a build.
+	ReasonDockerFileCreateFailed StepFailureReason = "DockerFileCreationFailed"
+
+	// ReasonInvalidArtifactsMappingFailed is the reason associated with an
+	// invalid artifacts mapping of files that need to be copied.
+	ReasonInvalidArtifactsMappingFailed StepFailureReason = "InvalidArtifactsMappingFailed"
+
+	// ReasonArtifactsFetchFailed is the reason associated with a failure to
+	// download specified scripts in the application image.
+	ReasonArtifactsFetchFailed StepFailureReason = "FetchScriptsFailed"
+
+	// ReasonFSOperationFailed is the reason associated with a failed fs
+	// operation. Create, remove directory, copy file, etc.
+	ReasonFSOperationFailed StepFailureReason = "FileSystemOperationFailed"
+
+	// ReasonInstallScriptsFailed is the reason associated with a failure to
+	// install scripts in the builder image.
+	ReasonInstallScriptsFailed StepFailureReason = "InstallScriptsFailed"
+
+	// ReasonGenericS2IBuildFailed is the reason associated with a broad range of
+	// failure.
+	ReasonGenericS2IBuildFailed StepFailureReason = "GenericS2IBuildFailed"
+
+	// ReasonUnmentDependenciesFailed is the failure reason associated with a builder
+	// image that doesn't contain required dependencies for building the app.
+	ReasonUnmentDependenciesFailed StepFailureReason = "UnmetBuilderImageDependenciesFailed"
```

Comment:
1) should be "unmet" 
2) it's not dependencies for building the app, it's really s2i dependencies, right?  (tar, sh)